README
 $Id: README,v 1.4.2.1 2011/05/18 07:33:21 ak Exp $
-----------------------------------------------------------------------
 Copyright (C) 2009-2011 Cubicroot Co. Ltd. ALL RIGHTS RESERVED.
 Developed by Cubicroot Co. Ltd. (http://cubicroot.jp/)
------------------------------------------------------------------------
 bounceHammer: An open source software mainly developed for handling 
               email bounces on UNIX/Linux server.

 Other information is available at the following sites:

  - http://sourceforge.jp/projects/bouncehammer/
  - http://twitter.com/bouncehammer
------------------------------------------------------------------------

ARTISTIC -------------- Artistic License
AUTHORS --------------- Author of bounceHammer
GNU-GPL --------------- GNU General Public License version 2
INSTALL --------------- How to install bounceHammer
Implement.PL ---------- Check -T, -B operation on file handle
KNOWNBUGS ------------- Known bugs in bounceHammer
LICENSE --------------- About license
Makefile.am ----------- Makefile
Makefile.in ----------- Makefile
Modules.PL* ----------- Check depended perl modules
README ---------------- This file
RELEASE_NOTES --------- Release history of bounceHammer(English)
RELEASE_NOTES.JA ------ Release history of bounceHammer(Japanese)
SUPPORT --------------- About bounceHammer commercial support
acinclude.m4 ---------- For autoconf
aclocal.m4 ------------
config.guess ----------
config.sub ------------
configure ------------- configure script
configure.ac ----------
diff/ ----------------- Patches for a certain environment
examples/ ------------- Examples for make test
install-sh ------------
missing ---------------
src/ ------------------ Source code
t/ -------------------- Test cases for make test
template/ ------------- WebUI template files


RELEASE_NOTES
 $Id: RELEASE_NOTES,v 1.76.2.22 2011/10/10 09:56:54 ak Exp $
 -Id: RELEASE_NOTES,v 1.12 2009/12/01 10:35:23 ak Exp -
 -Id: RELEASE_NOTES,v 1.34 2009/08/27 05:09:04 ak Exp -
-----------------------------------------------------------------------
 ##                                      ##  ##                                     
 ##      ####  ##  ## #####   #### ####  ##  ##  ####  ##  ## ##  ##  ####  #####   
 #####  ##  ## ##  ## ##  ## ##   ##  ## ######     ## ###### ###### ##  ## ##  ##  
 ##  ## ##  ## ##  ## ##  ## ##   ###### ##  ##  ##### ###### ###### ###### ##      
 ##  ## ##  ## ##  ## ##  ## ##   ##     ##  ## ##  ## ##  ## ##  ## ##     ##      
 #####   ####   ##### ##  ##  #### ####  ##  ##  ##### ##  ## ##  ##  ####  ##      

                   BOUNCEHAMMER/RELEASE_NOTES
-----------------------------------------------------------------------
 Copyright (C) 2009-2011 Cubicroot Co. Ltd. ALL RIGHTS RESERVED.
 Developed by Cubicroot Co. Ltd. (http://cubicroot.jp/)
-----------------------------------------------------------------------
 This listing show the release notes of "bounceHammer",  an open source
 software mainly developed for handling email bounces on UNIX server.
 The list shows the version of bounceHammer engine, the version of con-
 figuration file of bounceHammer, the date of release, and a summary of
 the changes in that release. More information about bounceHammer is a-
 vailable at our site: http://bouncehammer.jp/ .

 Other information is available at the following sites:

  - http://sourceforge.jp/projects/bouncehammer/
  - http://twitter.com/bouncehammer
  - http://www.facebook.com/bouncehammer
-----------------------------------------------------------------------

2.7.7/2.7.0	2011/11/14

  New functionality:
    * New command line option for bin/mailboxparser, --save-failed-mesg
      is for save messages which are failed to parse.

  Assorted improvements and code cleanup:
    * Update for commercial MTA modules.
    * Improved analytical accuracy for bounce messages via Exim, qmail,
      Sendmail, Postfix, and au by KDDI. Thanks to Toyokawa.


-----------------------------------------------------------------------
2.7.6/2.7.0	2011/10/03

  Assorted improvements and code cleanup:
    * @m.facebook.com is classified into Webmail/facebook .
    * @verizon.net is classified into Smartphone/verizon.

  Fixed bugs:
    * Fix the 3rd argument of AC_PATH_PROG() macro in ./configure .
    * Fixed an issue that  bin/mailboxparser stops  with the error mes-
      sage: ``Invalid mailbox list: ...'' if any ESC(0x1b) character is
      contained in a phrase part of a email header. Thanks to Toyokawa.


-----------------------------------------------------------------------
2.7.5/2.7.0	2011/08/25

  Assorted improvements and code cleanup:
    * Support a bounced email from Biglobe (****@***.biglobe.ne.jp) and
      Biglobe-compatible emails (tmtv.ne.jp,inacatv.ne.jp,ttv.ne.jp).
    * Add some webmail domains of qip.ru (RU::WebMail).
    * Add 81 webmail domains of Shtrudel M@il (US::WebMail).
    * MTA Module list on the WebUI.

  Fixed bugs:
    * Fixed csv format bug ( multiple header fields are displayed ) in
      Kanadzuchi/Log.pm .

  New files:
    * lib/Kanadzuchi/MTA/MODULES
    * lib/Kanadzuchi/MTA/JP/Biglobe.pm
    * lib/Kanadzuchi/MTA/User/Template.pm (FFR)
    * share/template/stylesheet/mtamodules.css
    * share/template/page/mtamodules.html
    * share/template/l10n/{ja,en}/help-mta-modules.html
    

-----------------------------------------------------------------------
2.7.4/2.7.0	2011/07/29

  Assorted improvements and code cleanup:
    * Support a bounced (blocked) email from @facebook.com and @groups.
      facebook.com.
    * WebUI template and CSS tiny improvement.

  Fixed bugs:
    * Fixed an issue that the following warning message: ``Use of unin-
      itialized value in pattern match (m//)  at lib/Kanadzuchi/MTA/US/
      Google.pm line 194.''.
    * Fixed an issue that could not parse  an email which the last line
      does not end with LF.
    * Fixed an issue that bin/mailboxparser cannot parse bounced emails
      in Maildir except the first email when bounced emails of the same
      pattern are continuous.  Thanks to REALIT Co., Ltd.

  New files:
    * lib/Kanadzuchi/MTA/US/Facebook.pm (TESTED RCP-P1 and RCP-P2 ONLY)


-----------------------------------------------------------------------
2.7.3/2.7.0	2011/06/23

  Fixed bugs:
    * Fixed  an issue  that  precision of analysis  for a bounced email
      from au by KDDI (@ezweb.ne.jp:  Japanese cellular phone provider)
      have dropped. Thanks to REALIT Co., Ltd.


-----------------------------------------------------------------------
2.7.2/2.7.0	2011/06/20

  Assorted improvements and code cleanup:
    * WebUI template and CSS tiny improvement.
    * Analytical accuracy of mailboxparser improvement for mailboxfull
      error from au by KDDI (Japanese cellular phone provider). Thanks
      to REALIT Co., Ltd.

  Fixed bugs:
    * Fixed an issue that libexec/mp-startup.pl does not work properry
      if bouncehammer was installed with --with-perl-extlib option.
    * Fixed an issue that ``Illegal division by zero at mailboxparser
      line 462.''


-----------------------------------------------------------------------
2.7.1/2.7.0	2011/05/25

  Assorted improvements and code cleanup:
    * New configure option --with-perl-extlib=dir, dir is the directory
      which contain required perl modules installed by cpanm -l.
    * Support My Opera Mail's domain: @myopera.com .
    * Update syslog format of each command.
    * Update email domains of fastmail.fm and Excite.
    * Update webmail domains of some countries in  Europe and Southeast
      Asia.
    * Support some smartphone domains of America and Indonesia.
    * A bounced email which is rejected by  AOL Phoenix's Sender Filter
      will be classified into the reason 'filtered',and by Content Fil-

src/lib/Kanadzuchi/Log.pm
# $Id: Log.pm,v 1.21.2.1 2011/08/19 09:03:59 ak Exp $
# -Id: Log.pm,v 1.2 2009/10/06 06:21:47 ak Exp -
# -Id: Log.pm,v 1.11 2009/07/16 09:05:33 ak Exp -
# Copyright (C) 2009,2010 Cubicroot Co. Ltd.
# Kanadzuchi::
                      
 ##                   
 ##     ####   #####  
 ##    ##  ## ##  ##  
 ##    ##  ## ##  ##  
 ##    ##  ##  #####  
 ###### ####      ##  
              #####   
package Kanadzuchi::Log;

#  ____ ____ ____ ____ ____ ____ ____ ____ ____ 
# ||L |||i |||b |||r |||a |||r |||i |||e |||s ||
# ||__|||__|||__|||__|||__|||__|||__|||__|||__||
# |/__\|/__\|/__\|/__\|/__\|/__\|/__\|/__\|/__\|
#
use base 'Class::Accessor::Fast::XS';
use strict;
use warnings;
use Kanadzuchi::Metadata;
use Time::Piece;

#  ____ ____ ____ ____ ____ ____ ____ ____ ____ 
# ||A |||c |||c |||e |||s |||s |||o |||r |||s ||
# ||__|||__|||__|||__|||__|||__|||__|||__|||__||
# |/__\|/__\|/__\|/__\|/__\|/__\|/__\|/__\|/__\|
#
__PACKAGE__->mk_accessors(
	'directory',	# (Path::Class::Dir) log directory
	'logfile',	# (Path::Class::File::Lockable) log file.
	'entities',	# (Ref->Array) K::M::* object
	'files',	# (Path::CLass::File) Temporary log files
	'count',	# (Integer) the number of bounced messages
	'format',	# (String) Log format
	'header',	# (Integer) 1 = Output the header part
	'footer',	# (Integer) 1 = Output the footer part
	'comment',	# (String) Additional description in the header
	'device',	# (String) Log device, file handle, screeen, ...
);

#  ____ ____ ____ ____ ____ ____ _________ ____ ____ ____ ____ 
# ||G |||l |||o |||b |||a |||l |||       |||v |||a |||r |||s ||
# ||__|||__|||__|||__|||__|||__|||_______|||__|||__|||__|||__||
# |/__\|/__\|/__\|/__\|/__\|/__\|/_______\|/__\|/__\|/__\|/__\|
#
my $OutputFormat = { 'yaml' => q(), 'json' => q(), 'csv' => q() };
my $OutputHeader = { 'yaml' => q(), 'json' => q(), 'csv' => q() };
my $RecDelimiter = { 'yaml' => q(), 'json' => ',', 'csv' => q() };

# Dump with YAML/JSON format
$OutputFormat->{'json'} .= qq|{ "bounced": %d, "addresser": "%s", "recipient": "%s", |;
$OutputFormat->{'json'} .= qq|"senderdomain": "%s", "destination": "%s", "reason": "%s", |;
$OutputFormat->{'json'} .= qq|"hostgroup": "%s", "provider": "%s", "frequency": %d, |;
$OutputFormat->{'json'} .= qq|"description": %s, "token": "%s" }|;
$OutputFormat->{'yaml'} .= qq|- |.$OutputFormat->{'json'};

# Dump with CSV format
$OutputFormat->{'csv'} .= qq|%d,%s,%s,%s,%s,%s,%s,%s,%d,%s,%s,%s,%s,%s|;
$OutputHeader->{'csv'} .= q|bounced,addresser,recipient,senderdomain,destination,reason,|;
$OutputHeader->{'csv'} .= q|hostgroup,provider,frequency,deliverystatus,timezoneoffset,|;
$OutputHeader->{'csv'} .= q|diagnosticcode,mta,token|.qq|\n|;

#  ____ ____ ____ ____ ____ _________ ____ ____ ____ ____ ____ ____ ____ 
# ||C |||l |||a |||s |||s |||       |||M |||e |||t |||h |||o |||d |||s ||
# ||__|||__|||__|||__|||__|||_______|||__|||__|||__|||__|||__|||__|||__||
# |/__\|/__\|/__\|/__\|/__\|/_______\|/__\|/__\|/__\|/__\|/__\|/__\|/__\|
#
sub new
{
	# +-+-+-+
	# |n|e|w|
	# +-+-+-+
	#
	# @Description	Wrapper method of new()
	# @Param
	# @Return	Kanadzuchi::Log Object
	my $class = shift();
	my $argvs = { @_ };

	DEFAULT_VALUES: {
		$argvs->{'format'} = 'yaml' unless( $argvs->{'format'} );
		$argvs->{'comment'} = q() unless( $argvs->{'comment'} );
	}
	return $class->SUPER::new( $argvs );
}

#  ____ ____ ____ ____ ____ ____ ____ ____ _________ ____ ____ ____ ____ ____ ____ ____ 
# ||I |||n |||s |||t |||a |||n |||c |||e |||       |||M |||e |||t |||h |||o |||d |||s ||
# ||__|||__|||__|||__|||__|||__|||__|||__|||_______|||__|||__|||__|||__|||__|||__|||__||
# |/__\|/__\|/__\|/__\|/__\|/__\|/__\|/__\|/_______\|/__\|/__\|/__\|/__\|/__\|/__\|/__\|
#
sub logger
{
	# +-+-+-+-+-+-+
	# |l|o|g|g|e|r|
	# +-+-+-+-+-+-+
	#
	# @Description	Log to the file(JSON::Syck::Dump)
	# @Param	<None>
	# @Return	0 = device not found or no record to log
	#		n = the number of logged records
	# @See		dumper()
	my $self = shift(); $self->{'format'} = q(yaml);
	my $reqv = defined(wantarray()) ? 1 : 0;
	my $data = undef();

	if( $reqv == 1 )
	{
		$data = $self->dumper();
		return $data;
	}
	else
	{
		return $self->dumper();
	}
}

sub dumper
{
	# +-+-+-+-+-+-+
	# |d|u|m|p|e|r|
	# +-+-+-+-+-+-+
	#
	# @Description	Dump to the file or screen with format
	# @Param	<None>
	# @Return	0 = No record to dump 
	#		1 = Successfully dumped
	my $self = shift();
	my $atab = undef();
	my $data = undef();
	my $reqv = defined(wantarray()) ? 1 : 0;
	my $damn = {};
	my $head = q();
	my $foot = q();

	return(0) if( $self->{'count'} == 0 );

	# Output header
	if( $self->{'format'} ne 'csv' )
	{
		my $_t = localtime();
		$OutputHeader->{ $self->{'format'} } = '# Generated: '.$_t->ymd('/').' '.$_t->hms(':').qq| \n|,
	}

	# Decide header and footer
	if( $self->{'header'} )
	{
		$head .= $OutputHeader->{ $self->{'format'} };
		$head .= q|# |.qq|$self->{'comment'}\n| if( length($self->{'comment'}) );
	}

	if( $self->{'footer'} )
	{
		$foot .= q|# |.qq|$self->{'comment'}\n| if( length($self->{'comment'}) );
	}

	# Print header
	if( $self->{'format'} eq 'asciitable' )
	{
		require Text::ASCIITable;
		$atab->{'tab'} = new Text::ASCIITable( { 'headingText' => 'Bounce Messages' } );
		$atab->{'tab'}->setOptions( 'outputWidth', 80 );
		$atab->{'tab'}->setCols( '#', 'Date', 'Addresser', 'Recipient', 'Stat', 'Reason' );
		$atab->{'num'} = 0;
	}
	else
	{
		$data .= $head;
	}

	# Print left square bracket character for the format JSON
	$data .= '[ ' if( $self->{'format'} eq q(json) );

	PREPARE_LOG: foreach my $_e ( @{$self->{'entities'}} )
	{
		$damn = $_e->damn();

		if( defined($atab) )
		{
			$atab->{'num'}++;
			$damn->{'datestring'} = $_e->bounced->ymd('/').' '.$_e->bounced->hms(':');
			$atab->{'tab'}->addRow( $atab->{'num'}, $damn->{'datestring'}, $damn->{'addresser'},
				$damn->{'recipient'}, $damn->{'deliverystatus'}, $damn->{'reason'} );
		}
		else
		{
			if( $self->{'format'} eq 'csv' )
			{
				$damn->{'diagnosticcode'} =~ y{,}{ };
				$data .= sprintf( $OutputFormat->{ $self->{'format'} },
						$damn->{'bounced'}, $damn->{'addresser'}, $damn->{'recipient'},
						$damn->{'senderdomain'}, $damn->{'destination'}, $damn->{'reason'}, 
						$damn->{'hostgroup'}, $damn->{'provider'}, $damn->{'frequency'}, 
						$damn->{'deliverystatus'}, $damn->{'timezoneoffset'},
						$damn->{'diagnosticcode'}, $damn->{'smtpagent'}, $damn->{'token'} );

			}
			else
			{
				$data .= sprintf( $OutputFormat->{ $self->{'format'} },
						$damn->{'bounced'}, $damn->{'addresser'}, $damn->{'recipient'},
						$damn->{'senderdomain'}, $damn->{'destination'},
						$damn->{'reason'}, $damn->{'hostgroup'}, $damn->{'provider'}, 
						$damn->{'frequency'}, $damn->{'description'}, $damn->{'token'} );
			}
			$data .= $RecDelimiter->{ $self->{'format'} }.qq(\n);
		}

	} # End of foreach() PREPARE_LOG:

	# Replace the ',' at the end of data with right square bracket for the format JSON
	$data =~ s{,\n\z}{ ]\n} if( $self->{'format'} eq 'json' );

	if( defined($atab) && $atab->{'num'} > 0 )
	{
		$atab->{'tab'}->addRowLine();
		$atab->{'tab'}->addRow( q{}, q{}, q{}, q{Total}, $self->{'count'} );
		$data = $atab->{'tab'}->draw();
	}
	else
	{
		$data .= $foot if( length($foot) );
	}

	if( $reqv == 1 )
	{
		# Return as a scalar(dumped data)
		return $data;
	}
	else
	{
		# Dumped data are not required, return true;
		if( ref($self->{'device'}) eq q|IO::File| )
		{
			print( {$self->{'device'}} $data );
		}
		else
		{
			print( STDOUT $data );
		}
		return(1);
	}
}

1;
__END__
